mysql-test/suite/rpl/t/rpl_semi_sync_shutdown_await_ack.inc

```diff
@@ -111,19 +108,80 @@ show status like 'Rpl_semi_sync_master_no_tx';
 --echo #
 --echo #-- Re-synchronize slaves with master and disable semi-sync
 
+if (`SELECT ($server_2_expect_row_count + $server_3_expect_row_count) < 2`)
+{
+--echo #-- FIXME: workaround for MDEV-28141, preventing errored replicas from
+--echo # killing their semi-sync connections
+# I.e. we can't create a new kill connection to the primary if we know that the
+# primary is shutting down for risk of Packets out of order error. So we wait
+# to hit a debug_sync point before the creation of the new kill_connection, and
+# don't progress until the primary has been shutdown, so no new connection can
+# be formed.
+# Note this is only needed in the error case (using corrupt_queue_event), as
+# the running io_thread will otherwise automatically detect that the primary
+# has shutdown before progressing to the cleanup of the io thread.
+}
+
+if (!$server_2_expect_row_count)
+{
+  --connection server_2
+  set debug_sync= "now wait_for at_semisync_kill_connection";
+  set debug_sync= "now signal continue_semisync_kill_connection";
+  --echo # Wait for debug_sync signal to have been received before issuing RESET
+  let $wait_condition= select count(*)=0 from information_schema.processlist where state like "debug sync point%";
+  source include/wait_condition.inc;
+  set debug_sync= "reset";
+}
+if (!$server_3_expect_row_count)
+{
+  --connection server_3
+  set debug_sync= "now wait_for at_semisync_kill_connection";
+  set debug_sync= "now signal continue_semisync_kill_connection";
+  --echo # Wait for debug_sync signal to have been received before issuing RESET
+  let $wait_condition= select count(*)=0 from information_schema.processlist where state like "debug sync point%";
+  source include/wait_condition.inc;
+  set debug_sync= "reset";
+}
+
 --echo #-- Stop slaves
 
 --connection server_2
---eval SET @@GLOBAL.debug_dbug= "$sav_server_2_dbug"
---eval SET @@GLOBAL.rpl_semi_sync_slave_enabled= 0
---let $rpl_only_running_threads= 1
-source include/stop_slave.inc;
+if (!$server_2_expect_row_count)
+{
+  # ER_SLAVE_RELAY_LOG_WRITE_FAILURE
+  --let $slave_io_errno=1595
+  --source include/wait_for_slave_io_error.inc
+  --source include/stop_slave_sql.inc
+}
+if ($server_2_expect_row_count)
+{
+  --echo #-- Waiting for IO thread to realize master shutdown and stop
+  --let $slave_io_errno=2003
+  --source include/wait_for_slave_io_error.inc
+  --source include/stop_slave_sql.inc
+  #--source include/stop_slave.inc
+}
```

Comment:
Here, is it really necessary to wait_for_slave_io_error?
If I understand correctly, you set the retry count low and wait for the retry count to give up and the IO THREAD to fail with reconnect error.
Isn't it possible to just --source include/stop_slave.inc ?
I fixed wait_for_slave_io_to_stop.inc to disregards an error left by the IO thread, since this can happen in a number of scenarios.

In general, I think it should be possible to just --source include/stop_slave_io.inc to stop the IO thread in most normal cases. If there's a special reason that this is not possible in this test, please add a comment explaining why this is so.
If there is a

Reply:
We could do `stop_slave.inc` now, but in `wait_for_slave_io_to_stop.inc` which is used, as you mention, your fix includes:

```
# Disabled, as IO errors are left behind when a normal STOP SLAVE interrupts
# the initial communication between the IO thread and the master (MDEV-32892).
#if (!$rpl_allow_error)
#{
#  --let $slave_error_param= Last_IO_Errno
#}
```

so once we re-enable it after MDEV-32892, the test would fail again. We could preempt that by setting `rpl_allow_error` now too, and it would save a few seconds from the test run as we don't have to wait for the disconnection. But I slightly prefer consistent test behavior over that. Though if you have a strong preference to use `stop_slave.inc`, I'm also okay with that.

Reply:
I don't understand, what do you mean with "consistent test behaviour"?

If we want to stop the slave, then using stop_slave.inc is the consistent way, and the stop_slave.inc script needs to be working so that it is able to stop the slave. If the script is not able to do that now or in the future, it needs to be fixed.

If it's expected that the IO thread may or may not have reached an error condition at the point where we stop it, that's fine.

Setting the IO thread to retry connections to a shut-down master until giving up seems to only add unnecessary complexity, I don't see how that adds value to the test?

Reply:
> If it's expected that the IO thread may or may not have reached an error condition at the point where we stop it, that's fine.

The "may or may not" part is the inconsistent behavior. As if something does go wrong, there would be a clear expectation being broken, rather than having to debug through different possible outcomes.

Though note `stop_slave.inc` works fine, no adjustment necessary there.

Reply:
@knielsen I updated the patch to use stop slave, though I kept the stop of the sql and io threads separate to be explicit in the intent that only the io thread can error.